src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,205 @@
     public T get(int index) {
+        if (index < 0 || index >= size()) {
+            throw new IndexOutOfBoundsException("Indeks jest poza zakresem listy: " + index);
+        }
+
+        Node<T> current = head;
+        int currentIndex = 0;
+
+        while (current != null && currentIndex < index) {
+            current = current.next;
+            currentIndex++;
+        }
+        if (current != null && currentIndex == index) {
+            return current.value;
+        }
         return null;
     }
 
     @Override
     public T set(T value, int index) {
+        if (index < 0 || index >= size()) {
+            throw new IndexOutOfBoundsException("Indeks jest poza zakresem listy: " + index);
+        }
+
+        Node<T> current = head;
+        int currentIndex = 0;
+
+        while (current != null && currentIndex < index) {
+            current = current.next;
+            currentIndex++;
+        }
+
+        if (current != null && currentIndex == index) {
+            T oldValue = current.value;
+            current.value = value;
+            return oldValue;
+        }
+
         return null;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0 || index >= size()) {
+            throw new IndexOutOfBoundsException("Invalid index: " + index);
+        }
+
+        Node<T> nodeToRemove = getNodeAtIndex(index);
+        T valueToRemove = nodeToRemove.value;
+        removeNode(nodeToRemove);
+        return valueToRemove;
     }
 
-    @Override
     public boolean remove(T object) {
+        if (object == null) {
+            return removeNull();
+        } else {
+            return removeNonNull(object);
+        }
+    }
+
+    private boolean removeNull() {
+        Node<T> current = head;
+
+        while (current != null) {
+            if (current.value == null) {
+                removeNode(current);
+                return true;
+            }
+            current = current.next;
+        }
+
+        return false;
+    }
+
+    private boolean removeNonNull(T object) {
+        Node<T> current = head;
+
+        while (current != null) {
+            if (object.equals(current.value)) {
+                removeNode(current);
+                return true;
+            }
+            current = current.next;
+        }
+
         return false;
     }
 
+    private void removeNode(Node<T> node) {
+        if (node.prev != null) {
+            node.prev.next = node.next;
+        } else {
+            head = node.next;
+        }
+
+        if (node.next != null) {
+            node.next.prev = node.prev;
+        } else {
+            tail = node.prev;
+        }
+    }
+
+    private Node<T> getNodeAtIndex(int index) {
+        if (index < 0 || index >= size()) {
+            throw new IndexOutOfBoundsException("Invalid index: " + index);
+        }
+
+        Node<T> current = head;
+        for (int i = 0; i < index; i++) {
+            current = current.next;
+        }
```

Comment:
[checklist](https://github.com/mate-academy/jv-linked-list/blob/master/checklist.md#dont-iterate-your-nodes-from-the-head-to-the-tail-or-vice-versa
![image](https://github.com/mate-academy/jv-linked-list/assets/49767049/ed4acc50-29e3-407b-aa5f-96b5a92c890a)
)